runtime/jcl/common/java_lang_invoke_VarHandle.c

```diff
@@ -111,7 +112,11 @@ Java_java_lang_invoke_FieldVarHandle_lookupField(JNIEnv *env, jobject handle, jc
 		goto _cleanup;
 	}
 
-	J9VMJAVALANGINVOKEVARHANDLE_SET_MODIFIERS(vmThread, J9_JNI_UNWRAP_REFERENCE(handle), ((J9ROMFieldShape*) romField)->modifiers);
+	varHandle = J9_JNI_UNWRAP_REFERENCE(handle);
+	J9VMJAVALANGINVOKEVARHANDLE_SET_MODIFIERS(vmThread, varHandle, ((J9ROMFieldShape *)romField)->modifiers);
+	if (isStatic && (NULL != definingClass)) {
```

Comment:
It shouldn't be possible to get here with `isStatic` true and `definingClass == null`.  It would be better to assert that then let a partially initialized VH escape

Reply:
`StaticFieldVarHandle` and `InstanceFieldVarHandle` extend `FieldVarHandle`. For `InstanceFieldVarHandle`, `isStatic` will be `false`. Updating `definingClass` is not required for `InstanceFieldVarHandle` since `definingClass` is unused by `InstanceFieldVarHandle`. `definingClass` is only used by `StaticFieldVarHandle`.

We can add an assert for `definingClass` since I expect it to be always non-null for `StaticFieldVarHandle`.

In `FieldVarHandle`'s constructor, we set `definingClass` to `j9LookupClass`. A more efficient check would be:
```
if (isStatic && (j9LookupClass != definingClass)) {
    Assert_JCL_notNull(definingClass);
    ...
```

I also plan to add a new test case to address the failure.

Reply:
I was thinking:
```
if (isStatic) {
 Assert_JCL_notNull(definingClass);
}
```
Why add the other check?

Reply:
> Why add the other check?

Because we already set `FieldVarHandle.definingClass` to `lookupClass` in `FieldVarHandle`'s constructor before invoking the native `FieldVarHandle_lookupField` method. We can avoid setting `definingClass` the same value twice.

https://github.com/eclipse/openj9/blob/6a8a0aef2e12405cc08654f15c62952d7c5f4fc6/jcl/src/java.base/share/classes/java/lang/invoke/FieldVarHandle.java#L53-L60

Reply:
I hadn't realized we're hammering a final field in the native that had already been set.  I really don't like doing that.

We should probably remove the `final` modifier in this case.

Reply:
In the latest changes, `definingClass` is only updated once in the native method. Can we keep the `final` keyword?

Reply:
Even though, it is initialized once in native code. Compiler expects it to be initialized in Java code. So, the final keyword would need to be removed.